### PR TITLE
Add hocon support

### LINKED ISSUE #2: More configuration format support
Some formats to consider, and likely corresponding parsing libraries:

* XML - [Jackson](https://github.com/FasterXML/jackson-dataformat-xml)
* HOCON - [Lightbend](https://github.com/lightbend/config)
* Hjson - [hjson-java](https://github.com/hjson/hjson-java)
* TOML - ?
* INI - [inieditor-java](https://github.com/nikhaldi/inieditor-java)

### PR DESCRIPTION
**Pull request by CDFN**
Partially resolves #2. Still we're missing comment support. Also there's weird bug -
```java
public interface Config {
  @DefaultString("someString")
  String testString();
}
```
after writing to file results in
```conf
{
    # hardcoded value
    "testString" : "someString"
}
```
even though the comment was never created by me. Help with troubleshooting would be appreciated :)
Addonationaly, better exception handling. Currently in HoconConfigurationFactory#loadMapFromReader only ConfigException.Parse is handled, otherwise IOException is thrown.

### CODE REVIEW

**Review by A248 — COMMENTED** [review]
With regards to exception handling, AbstractConfigurationFactory#loadMapFromReader can be loosened to throw InvalidConfigException. Then you'll be able to better map HOCON's exceptions.

One more thing, have you tested nested configuration sections? I'm not sure ConfigFactory#parseMap is the right call, but I believe it is ConfigValueFactory#fromMap. ConfigValueFactory says its 'fromAnyRef' (which fromMap calls) is the inverse operation of ConfigObject#unwrapped.

On the decision to expose ConfigParseOptions, could you elaborate on what this would be used for? Some of the options in ConfigParseOptions do not seem pertinent to DazzleConf, though I'll admit I'm a little hazy on their purpose, so I may need to look closer into this.

I'll also be available to provide advice on implementing comment support. First, let's iron out these details.

**Comment by A248** [comment]
While looking into comment support, I found out where "hardcoded value" comes from: https://github.com/lightbend/config/blob/ea45ea3767a201933eeeb9c3f0f13eacc9b51f07/config/src/main/java/com/typesafe/config/impl/ConfigImpl.java#L156

**Comment by A248** [comment]
With that in mind, I understand how to add comment support:
1. Convert DazzleConf's Map<String, Object> to lightbend-config's Map<String, ConfigValue>. ConfigValueFactory will assist in this.
2. When converting the map values (Object -> ConfigValue), add comments with ConfigValue#withOrigin and ConfigOrigin#withComments
3. Render the complete ConfigValue which is derived from the Map<String, ConfigValue>.

**Comment by CDFN** [comment]
Thank you for support with implementing it. I haven't been testing nested sections yet. I decided to expose ParseOptions to make parsing config more flexible. Should I wrap underlying options in ParseOptions into HoconOptions so there's no need to use hocon lib's objects? Also thanks for explaining how should I implement comments. I'll try my best 😃

**Comment by A248** [comment]
On further thought, I think you are right to expose the formatting library's own options in full, as this is what is done by the other format extensions (Gson and SnakeYaml). However, I think ConfigRenderOptions should have saner defaults.

* *Origin* comments being enabled (`setOriginComments`) is what causes "hardcoded value" to show up, so origin comments should be disabled by default.
* `setJson` would be best set to false by default. It somewhat defeats the point of using hocon over json otherwise.

Let me know if you disagree with my reasoning on these. I'm not an everyday HOCON user so you likely know more than I do.

Also, feel free to add yourself to the <contributors> section in the parent pom.xml

File: hocon/src/main/java/space/arim/dazzleconf/ext/hocon/HoconOptions.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    Charset charset = StandardCharsets.UTF_8;
+
+    /**
+     * Sets parse options. Reading config uses them.
+     *
+     * @param options the options
+     * @return this builder
+     */
+    public Builder configParseOptions(ConfigParseOptions options) {
+      this.configParseOptions = options;
```
Comment (A248) [inline hocon/src/main/java/space/arim/dazzleconf/ext/hocon/HoconOptions.java:67]: This method, and the other builder methods, should use Objects.requireNonNull to validate their arguments.

File: hocon/src/main/java/space/arim/dazzleconf/ext/hocon/HoconOptions.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
+package space.arim.dazzleconf.ext.hocon;
+
+import com.typesafe.config.ConfigParseOptions;
+import com.typesafe.config.ConfigRenderOptions;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+public class HoconOptions {
```
Comment (A248) [inline hocon/src/main/java/space/arim/dazzleconf/ext/hocon/HoconOptions.java:9]:
```suggestion
public final class HoconOptions {
```

File: hocon/src/main/java/space/arim/dazzleconf/ext/hocon/HoconOptions.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  public Charset charset() {
+    return charset;
+  }
+
+  /**
+   * Builder of {@code HoconOptions}
+   *
+   * @author CDFN
+   */
+  public static class Builder {
```
Comment (A248) [inline hocon/src/main/java/space/arim/dazzleconf/ext/hocon/HoconOptions.java:52]:
```suggestion
  public static final class Builder {
```

**Review by A248 — COMMENTED** [review]
For new configuration formats, it would be better to encapsulate the identity of the ConfigurationFactory implementations, using static `create` methods rather than constructors.

The tomlj-support branch has an example of this:
https://github.com/A248/DazzleConf/blob/toml-support/tomlj/src/main/java/space/arim/dazzleconf/ext/tomlj/TomljConfigurationFactory.java

**Review by A248 — APPROVED** [review]
I took care of the rest. Thanks for contributing!

**Comment by CDFN** [comment]
Thanks for finishing it. Glad I could help even a little 🙂👍

**Pull request merged**